oap-server/server-alarm-plugin/src/main/java/org/apache/skywalking/oap/server/core/alarm/provider/expression/Expression.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.apache.skywalking.oap.server.core.alarm.provider.expression;
+
+import lombok.extern.slf4j.Slf4j;
+import org.mvel2.MVEL;
+import org.mvel2.ParserContext;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Expression support eval java basic expressions, just like groovy script
+ * The internal detail is it first compile the expression to a pareTree then execute the parseTree with data
+ * It caches the compiled expression for sake of performance
+ */
+@Slf4j
+public class Expression {
+    private Map<String, Object> expressionCache;
+    private ExpressionContext context;
+
+    public Expression(ExpressionContext context) {
+        this.context = context;
+        this.expressionCache = new ConcurrentHashMap<>();
+    }
+
+    /**
+     * Eval the given expression using empty data context
+     */
+    public Object eval(String expression) {
+        return eval(expression, null);
+    }
+
+    /**
+     * Eval the given expression with data context
+     */
+    public Object eval(String expression, Map<String, Object> vars) {
+        try {
+            Object obj = compile(expression, context);
+            return MVEL.executeExpression(obj, vars);
+        } catch (Throwable e) {
+            log.error("eval expression {} error", expression, e);
+            return null;
+        }
+    }
+
+    /**
+     * Compile the given expression to a parseTree
+     */
+    public Object compile(String expression, ExpressionContext pctx) {
+        Object o = expressionCache.get(expression);
+        if (o == null) {
+            o = MVEL.compileExpression(expression, pctx.getContext());
+            expressionCache.put(expression, o);
+        }
+        return o;
```

Comment:
Composite operation is not atomic, replace with `ConcurrentHashMap#computeIfAbsent`

Reply:
done

Reply:
I have a little doubt about this. What concurrency case makes us need `atomic`? @kezhenxu94

Reply:
@wu-sheng I didn't check the real case, but since @xbkaishui he is using `ConcurrentHashMap`, if there is no such case, a simple `HashMap` is enough

Reply:
OK, let's keep the `ConcurrentHashMap`, as it doesn't cost much. From my understanding, there is no concurrency or race condition.

Anyway, no big issue.